### PR TITLE
fix: make secret descriptor validation more granular (#95)

### PR DESCRIPTION
**Pull request by ThirdEyeSqueegee**
*Issue #, if available:*
#95 

*Description of changes:*
- Make secret descriptor validation more granular to account for edge cases as in #95 
- Add tests for various combinations of conflicting `objectName`s, `objectAlias`es, and `objectVersionLabel`s 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: provider/secret_descriptor.go [hunk 1]
```diff
@@ -337,29 +345,70 @@ func NewSecretDescriptorList(mountDir, translate, objectSpec string, regions []s
...
+		// -------------------------------------------
+		// | # | OBJECT ALIAS | OBJECT VERSION LABEL |
+		// |---|--------------|----------------------|
+		// | 1 | duplicate    | empty                |
+		// | 2 | empty        | duplicate            |
+		// | 3 | duplicate    | duplicate            |
+		// | 4 | empty        | empty                |
+		// -------------------------------------------
+		found, ok := seenNames[descriptor.ObjectName]
+		if ok {
```
Comment (simonmarty) [inline provider/secret_descriptor.go:359]: Use guard causes to remove nesting? (exit early)
```
if !ok {
// exit block
}

if !descHasAlias || !foundHasAlias {
// exit
}

...
```

File: provider/secret_descriptor.go [hunk 2]
```diff
@@ -337,29 +345,70 @@ func NewSecretDescriptorList(mountDir, translate, objectSpec string, regions []s
...
 
 		if len(descriptor.JMESPath) == 0 { //jmesPath not used. No more checks
 			continue
 		}
 
 		for _, jmesPathEntry := range descriptor.JMESPath {
-			if names[jmesPathEntry.ObjectAlias] {
-				return nil, fmt.Errorf("Name already in use for objectAlias: %s", jmesPathEntry.ObjectAlias)
+			if seenAliases[jmesPathEntry.ObjectAlias] {
+				return nil, fmt.Errorf("found duplicate object alias %s in JMES path entry %s", jmesPathEntry.ObjectAlias, jmesPathEntry.Path)
```
Comment (simonmarty) [inline provider/secret_descriptor.go:409]: I'm a bit concerned about this breaking customers that search for this error message, should we keep the start of the string the same?

**Review by simonmarty — CHANGES_REQUESTED** [review]

**Comment by simonmarty** [comment]
Could you also update the PR title to something more descriptive of the changes we're making?

File: provider/secret_descriptor.go (hunk 1 above)
Comment (ThirdEyeSqueegee) [reply provider/secret_descriptor.go:359]: Inverting the `if`s would honestly be messier since there's `else` cases for almost everything too and we want to run the rest of the validation logic after the `if ok` block; I think we should leave it as is unless there's a strong reason to invert (personally I don't think there's enough levels of nesting to necessitate inverting)

**Comment by dnlopes** [comment]
I'm _very_ interested in having this PR merged and added to the next release, for the same reasons the original issue was reported.

So, let me know if anything is needed, I can try to help.

**Comment by dnlopes** [comment]
Just a quick question on the behavior after this change:

by putting together both `AWSPREVIOUS` and `AWSCURRENT`, does this mean that both secrets are refreshed at the same time?

File: provider/secret_descriptor.go (hunk 1 above)
Comment (simonmarty) [reply provider/secret_descriptor.go:359]: Something like

```
	if ok {
			descHasAlias := descriptor.ObjectAlias != ""
			foundHasAlias := found.ObjectAlias != ""
			descHasVersionLabel := descriptor.ObjectVersionLabel != ""
			foundHasVersionLabel := found.ObjectVersionLabel != ""

			errorPrefix := fmt.Errorf("found descriptor with duplicate object name %s", descriptor.ObjectName)

			if descHasAlias && foundHasAlias && descriptor.ObjectAlias == found.ObjectAlias {
				if !descHasVersionLabel || !foundHasVersionLabel {
					// Case 1
					return nil, fmt.Errorf("%s, object alias %s with no version label", errorPrefix, descriptor.ObjectAlias)
				}
				
				if descriptor.ObjectVersionLabel == found.ObjectVersionLabel {
						// Case 3
					return nil, fmt.Errorf("%s, object alias %s, and version label %s", errorPrefix, descriptor.ObjectAlias, descriptor.ObjectVersionLabel)
				}
			} else {
				if !descHasVersionLabel || !foundHasVersionLabel {
					// Case 4
					return nil, fmt.Errorf("%s with no object alias or version label", errorPrefix)
				}

				if descriptor.ObjectVersionLabel == found.ObjectVersionLabel {
					// Case 2
					return nil, fmt.Errorf("%s and version label %s with no object alias", errorPrefix, descriptor.ObjectVersionLabel)
				}
			}
		}
```

There might be a way to denest this one step further even, it's a matter of finding which cases cause a return an short circuit those as early as possible in the code

It makes your code easier to test because you can clearly tell all the failure modes since they become sequential and they come up early in the subroutine

Comment (simonmarty) [reply provider/secret_descriptor.go:359]: Also, nit: maybe `ok` can be renamed to something more descriptive?

**Review by simonmarty — CHANGES_REQUESTED** [review]

File: provider/secret_descriptor.go (hunk 1 above)
Comment (ThirdEyeSqueegee) [reply provider/secret_descriptor.go:359]: Thanks, I totally blanked on the else-after-returns. And yeah I can rename `ok`, only reason I used it is that it appears to be the conventional name for the second variable returned during lookups

Comment (simonmarty) [reply provider/secret_descriptor.go:359]: If it's idiomatic you can keep it.

File: provider/secret_descriptor_test.go [hunk 3]
```diff
@@ -150,13 +151,148 @@ func TestConflictingName(t *testing.T) {
...
+          objectType: ssmparameter`
+
+	_, err := NewSecretDescriptorList("/", "", objects, singleRegion)
+
+	if err != nil {
+		t.Fatalf("Expected no error, got error: %v", err)
+	}
+}
+
+// Conflicting name; version label and alias present -- should not throw
```
Comment (simonmarty) [inline provider/secret_descriptor_test.go:276]: Could you update the test comments to help identify which "Cases" above each one of these matches to? That's gonna make future referencing easier.

Comment (ThirdEyeSqueegee) [reply provider/secret_descriptor_test.go:276]: Good call, will do

File: provider/secret_descriptor_test.go [hunk 4]
```diff
@@ -584,7 +720,7 @@ func TestVersionIdsMustMatch(t *testing.T) {
     - objectName: "MySecret1"
       objectType: ssmparameter
       objectVersion:  OldVersionId
-      failoverObject: 
+      failoverObject:
```
[the remark this answers is not in the document]
Comment (ThirdEyeSqueegee) [reply provider/secret_descriptor_test.go:723]: Sure thing

[the remark this answers is not in the document]
Comment (ThirdEyeSqueegee) [reply provider/secret_descriptor_test.go:744]: There was already a workflow step to check formatting but I don't think it was working, swapped it out for an action that checks formatting using `gofmt` in #446

**Comment by dnlopes** [comment]
Hello all @simonmarty @ThirdEyeSqueegee, first of all thanks for this PR, it will address a blocking issue with one of our use cases.

What can be done to move this PR forward? I'm willing to help if needed, but it seems everything is mostly done.

What is missing? Just bandwidth from the codeowners to review this?

Thanks in advance!

File: provider/secret_descriptor_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonmarty) [reply provider/secret_descriptor_test.go:744]: Spoke offline, done in https://github.com/aws/secrets-store-csi-driver-provider-aws/pull/446

**Review by simonmarty — APPROVED** [review]

**Review by gnaikrah — APPROVED** [review]
